// mcp_control/PageControl.js
var dao = require('mcp_dao');
var dc = dao.dc;

var esut = require('easy_util');
var digestUtil = esut.digestUtil;

var adminPageControl = require("./AdminPageControl.js");
var termPageCtrl = require('./TermPageControl.js');
var gamePageCtrl = require('./GamePageControl.js');
var stationPageCtrl = require('./StationPageControl.js');
var ticketPageCtrl = require('./TicketPageControl.js');
var orderPageCtrl = require('./OrderPageControl.js');
var moneyLogPageCtrl = require('./MoneyLogPageControl.js');
var printPageCtrl = require('./PrintPageControl.js');

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    console.log(bodyNode);
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[0]](headNode, bodyNode, cb);
}


PageControl.prototype.admin = function(headNode, bodyNode, cb)
{
    adminPageControl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.term = function(headNode, bodyNode, cb)
{
    termPageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.game = function(headNode, bodyNode, cb)
{
    gamePageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.station = function(headNode, bodyNode, cb)
{
    stationPageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.ticket = function(headNode, bodyNode, cb)
{
    ticketPageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.order = function(headNode, bodyNode, cb)
{
    orderPageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.moneylog = function(headNode, bodyNode, cb)
{
    moneyLogPageCtrl.handle(headNode, bodyNode, cb);
}

PageControl.prototype.print = function(headNode, bodyNode, cb)
{
    printPageCtrl.handle(headNode, bodyNode, cb);
}

var pageControl = new PageControl();
module.exports = pageControl;
